ubi_manifest/worker/tasks/depsolver/modulemd_depsolver.py

```diff
@@ -0,0 +1,159 @@
+
+MAX_WORKERS = int(os.getenv("UBI_MANIFEST_DEPSOLVER_WORKERS", "8"))
+
+
+class ModularDepsolver:
+    """
+    Class for depsolving modulemd units
+    """
+
+    def __init__(
+        self,
+        modular_items: List[ModularDepsolverItem],
+        input_repos: List[YumRepository],
+    ) -> None:
+        self._modular_items: List[ModularDepsolverItem] = modular_items
+        self._input_repos: List[YumRepository] = input_repos
+        # executor for this class, not adding retries because for pulp
+        # we use executor from pulplib
+        self._executor = Executors.thread_pool(max_workers=MAX_WORKERS)
+
+        # set of all already searched modules to avoid duplication & cycles
+        self._searched_modules: Set[str] = set()
+        # output set of resolved modulemd packages
+        self.modules: List[UbiUnit] = []
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        self._executor.__exit__(*args, **kwargs)
+
+    def run(self):
+        """
+        Run depsolver for each moudular dependency - recursively resolve all of
+        its modular dependencies and add binary and debug dependencies to list.
+        """
+        for item in self._modular_items:
+            modulemds_criteria = _get_criteria_for_modules(item.modulelist)
+            for module in item.modulelist:
+                self._update_searched_modules(module)
+            modules = f_proxy(
+                self._executor.submit(
+                    search_modulemds, modulemds_criteria, item.in_pulp_repos
+                )
+            )
+            # recurrently resolve dependencies for found modules
+            self._depsolve_modules(modules)
+        return self
+
+    def _depsolve_modules(self, modules):
+        """
+        Update modulemd output set with latest versions of modules, then find
+        dependencies for the modules and depsolve them.
+        """
+        filtered_modules = _get_modulemd_output_set(modules)
+        self.modules.extend(filtered_modules)
+
+        for module in filtered_modules:
+            # If dependencies is None, skip it
+            if not module.dependencies:
+                continue
+            # Get all dependencies and if not already processed, depsolve them as well
+            for dependency in module.dependencies:
+                if self._not_already_searched(dependency):
+                    self._update_searched_modules(dependency)
+                    modulemds_criteria = _get_criteria_for_modules(
+                        [
+                            dependency,
+                        ]
+                    )
+                    new_modules = f_proxy(
+                        self._executor.submit(
+                            search_modulemds, modulemds_criteria, self._input_repos
+                        )
+                    )
+                    self._depsolve_modules(new_modules)
+
+    def _update_searched_modules(self, module):
```

Comment:
I don't that one-line code update requires dedicated method. (minor)

Reply:
I have had to modify the function to reflect searching for all the streams of modulemds. Since the function is now a bit bigger, I kept it there. I can remove the function if appropriate.